### PR TITLE
[Object API] add object provider search

### PR DESCRIPTION
**Pull request by davetsay**
## Overview
Addresses #3609 

## Testing Instructions 
see #3609 

## Author Checklist
| | |
| --- | :---: |
| Changes address original issue? | Y |
| Unit tests included and/or updated with changes? | Not Yet |
| Command line build passes? | Y |
| Changes have been smoke-tested? | Y |
| Testing instructions included? | Y |

### CODE REVIEW

**Comment by davetsay** [comment]
@akhenry , wondering if we need to worry about result aggregation for object api search? sorting? Looking at the current searchAggregator.

**Comment by akhenry** [comment]
@davetsay  Looks good so far. We definitely want to do result aggregation. Sorting might be tricky because of what I'm about to suggest next...

I am wondering whether search might be a good candidate for a generator function? That way we could yield results as they come in, rather than blocking until the slowest search finishes. We might even want to extend this down to the object providers, so that they can yield results as soon as they are available (some search implementations may support this, I don't know). Obviously, a generator function would make sorting trickier (but not impossible). The sort would need to happen in real time in the tree, rather than the search results themselves. Maintaining sort in the tree is actually a capability we have talked about for a while, so we would be killing two avians with one aggregate of mineral or mineraloid matter.

**Comment by davetsay** [comment]
> so we would be killing two avians with one aggregate of mineral or mineraloid matter.

Didn't see this until now because we chatted about your comment. Perfection.

File: src/api/objects/ObjectAPI.js [hunk 1]
```diff
@@ -189,6 +189,16 @@ define([
         });
     };
 
+    ObjectAPI.prototype.search = function* (options) {
+        for (const provider of Object.values(this.providers)) {
+            if (provider.search !== undefined) {
+                yield provider.search(options);
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/api/objects/ObjectAPI.js:195]: So I think our search providers will also need to be able to yield in order to support search backends that either stream search results or return them in pages (we may hit a page limit in YAMCS for example). 

I'm a generator function n00b, but it looks as though generator functions can return generator functions. I haven't dug too deeply into the practical implications of this though. Basically we want the `.search` function to be able to yield arrays of objects from search providers before they've finished searching, if they support it.

If we can make this optional (and not force generator functions on plugins unless they want to use them) then great, otherwise no big deal if it's mandatory.

File: src/api/objects/ObjectAPI.js [hunk 2]
```diff
@@ -189,6 +189,16 @@ define([
         });
     };
 
+    ObjectAPI.prototype.search = function* (options) {
```
Comment (akhenry) [inline src/api/objects/ObjectAPI.js:192]: We need some jsdoc here to describe how this works both for callers, and for search providers.

File: src/api/objects/ObjectAPI.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (akhenry) [reply src/api/objects/ObjectAPI.js:195]: Also, now that I've thought about this a little more, this is going to block until the first provider yields something before it even queries subsequent providers right? That's not what we want.

Is there an approach that will allow us yield results from any of the search providers as soon as they're ready? I'm way into unknown territory here...Maybe generator functions aren't the right way of handling this, maybe we should just use event emitters?

File: src/plugins/condition/components/inspector/ConditionSetSelectorDialog.vue [hunk 3]
```diff
@@ -100,8 +101,22 @@ export default {
             selectedItem: undefined
         };
     },
+    computed: {
+        shouldDisplayNoResultsText() {
```
Comment (akhenry) [inline src/plugins/condition/components/inspector/ConditionSetSelectorDialog.vue:105]: Nice 👌 🎷

File: src/api/objects/ObjectAPI.js [hunk 4]
```diff
@@ -189,6 +189,16 @@ define([
...
+        for (const provider of Object.values(this.providers)) {
+            if (provider.search !== undefined) {
+                yield provider.search(options);
+            }
+        }
+
+        yield this.fallbackProvider.superSecretFallbackSearch(options);
```
Comment (akhenry) [inline src/api/objects/ObjectAPI.js:199]: We should convert these to new style domain objects so we're not handling a mix of legacy and new style domain objects in the tree

**Review by akhenry — COMMENTED** [review]
This is looking awesome. Some questions inline, some of which stem from my ignorance of generator functions.

File: src/api/objects/ObjectAPI.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (davetsay) [reply src/api/objects/ObjectAPI.js]: As we discussed, the generator yields promises so the yield values are ready immediately. The unit tests show object provider searches are run individually and non-blocking.

I do have a slight concern and might have to go back at some point to address handling network errors and such.

File: src/api/objects/ObjectAPI.js (hunk 2 above)
Comment (davetsay) [reply src/api/objects/ObjectAPI.js]: Added jsdoc

File: src/api/objects/ObjectAPI.js (hunk 4 above)
Comment (davetsay) [reply src/api/objects/ObjectAPI.js]: good call. done

**Comment by davetsay** [comment]
@akhenry , addressed review comments. Ready for re-review.

**Comment by akhenry** [comment]
So, I don't think generators are actually contributing anything here. We could achieve exactly the same thing with Promises by removing the yield statements and generator functions. This is my fault, I proposed using generators for this without really appreciating how complex the task actually is with generators.

What I had hoped is that we could use generators to allow the individual search providers to yield partial search results, as soon as they are available (eg. here's my first 10 results, but I'm still searching! I'll let you know when I'm done). Our search API would then yield partial results from any of the providers __as soon as they yield them__.

While this appears to be technically possible ([there is a stack overflow on this topic](https://stackoverflow.com/questions/50585456/how-can-i-interleave-merge-async-iterables)) it looks extremely convoluted. So, I've come to the conclusion that this is a premature optimization. One day our API will allow search providers to yield partial results as soon as they're available, but today is not that day.

**Comment by davetsay** [comment]
> So, I don't think generators are actually contributing anything here. We could achieve exactly the same thing with Promises by removing the yield statements and generator functions. This is my fault, I proposed using generators for this without really appreciating how complex the task actually is with generators.
> 
> What I had hoped is that we could use generators to allow the individual search providers to yield partial search results, as soon as they are available (eg. here's my first 10 results, but I'm still searching! I'll let you know when I'm done). Our search API would then yield partial results from any of the providers **as soon as they yield them**.
> 
> While this appears to be technically possible ([there is a stack overflow on this topic](https://stackoverflow.com/questions/50585456/how-can-i-interleave-merge-async-iterables)) it looks extremely convoluted. So, I've come to the conclusion that this is a premature optimization. One day our API will allow search providers to yield partial results as soon as they're available, but today is not that day.
> 
> TL;DR we should remove the generators, they're not doing anything.

Le Sigh... yeah...

@akhenry, made the changes. Updated tests as well and smoke tested in viper local environment.

image

**Review by akhenry — APPROVED** [review]
Looks great, thanks!

**Comment by akhenry** [comment]
Well, in this case I was wrong about using generators, but thanks anyway Barf.

**Comment by akhenry** [comment]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue) Y
